### DISCUSSION TITLE
WN 0007 - Nested Computations

### DISCUSSION THREAD

**Discussion by lloydtabb** [malloydata/malloy#1428] [What's Next: Language Proposals]
Malloy produces nested data.  To access nested data today, malloy implicitly unnests while guaranteeing computational accuracy.  There are times when we'd like to make computations against nested data without unnesting (count the number of occurrences with a filter) or to transform the nested data or to unnest the data more than once (to find co-occurrence).

https://github.com/malloydata/whatsnext/blob/main/wns/WN-0007/wn-0007.md

**Comment by christopherswenson** [comment]
I think pretty much all of this sounds good to me, except the exact syntax. I don't like making "return" a special name — that would mean that if you had a table with a column called `return`, you might run into some problems.

Two possible other syntaxes (I like # 1 better)

1. `return aggregate:`

`aggregate:`, `group_by:`, and `calculate:` can be prefixed by `return` to mark them as the return value.

I imagine we'll start adding other prefixes like `private` later, so this is probably consistent with future features.

Value could be named or unnamed.

```
dimension: foo is some_join -> {
  return aggregate: count()
}
```

2. `return:`

`return` would accept a dimension or measure, named or unnamed.

Downsides:
- Not clear to a reader whether the return value is an aggregate or scalar
- No way to use `calculate`s as a return, because you can't get at output space values

```
dimension: foo is some_join -> {
  return: count()
}
```

**Comment by christopherswenson** [comment]
It seems like the [Nested Sources](https://github.com/malloydata/whatsnext/blob/main/wns/WN-0007/wn-0007.md#nested-sources) section should be its own WN?
